### PR TITLE
PushSalesforceCourseStats fixes

### PR DESCRIPTION
**Pull request by Dantemss**
Reduce memory usage of PushSalesforceCourseStats
Don't send courses without a SF Contact ID
Fix sending preview courses to SF

Based on v52.0.1 so we can tag this for release

### CODE REVIEW

File: app/routines/push_salesforce_course_stats.rb [hunk 1]
```diff
@@ -42,7 +42,9 @@ def exec
 
   def applicable_courses
     # Don't update courses that have ended
-    terms = CourseProfile::Models::Course.terms.values_at(:spring, :summer, :fall, :winter)
+    terms = CourseProfile::Models::Course.terms.values_at(
+      :spring, :summer, :fall, :winter, :preview
```
Comment (Dantemss) [inline app/routines/push_salesforce_course_stats.rb:46]: Preview course fix

File: app/routines/push_salesforce_course_stats.rb [hunk 2]
```diff
@@ -52,41 +54,59 @@ def applicable_courses
...
-              "course_profile_courses"."id"
+            INNER JOIN "entity_roles"
+              ON "entity_roles"."id" = "course_membership_teachers"."entity_role_id"
+            INNER JOIN "user_profiles"
+              ON "user_profiles"."id" = "entity_roles"."user_profile_id"
+            INNER JOIN "openstax_accounts_accounts"
+              ON "openstax_accounts_accounts"."id" = "user_profiles"."account_id"
+            WHERE "openstax_accounts_accounts"."salesforce_contact_id" IS NOT NULL
+              AND "course_membership_teachers"."course_profile_course_id" =
+                "course_profile_courses"."id"
```
Comment (Dantemss) [inline app/routines/push_salesforce_course_stats.rb:65]: Don't send courses without any contact ID

File: app/routines/push_salesforce_course_stats.rb [hunk 3]
```diff
@@ -52,41 +54,59 @@ def applicable_courses
...
         WHERE_SQL
       )
 
     courses = courses.where(is_preview: false).or(courses.where.not(preview_claimed_at: nil))
 
-    courses.preload(
-      :offering, periods: { students: { role: { taskings: :task } } },
-                 teachers: { role: { profile: :account } }
-    )
+    courses.preload :offering, :periods, teachers: { role: { profile: :account } }
```
Comment (Dantemss) [inline app/routines/push_salesforce_course_stats.rb:72]: Reduce memory usage

File: app/routines/push_salesforce_course_stats.rb [hunk 4]
```diff
@@ -52,41 +54,59 @@ def applicable_courses
...
+      :course_membership_period_id, :is_comped, :is_paid, :first_paid_at, :dropped_at, <<~SQL
+        (
+          SELECT COALESCE(SUM("tasks_tasks"."completed_steps_count"), 0)
+          FROM "entity_roles"
+            INNER JOIN "tasks_taskings" ON "tasks_taskings"."entity_role_id" = "entity_roles"."id"
+            INNER JOIN "tasks_tasks" ON "tasks_tasks"."id" = "tasks_taskings"."tasks_task_id"
+          WHERE "entity_roles"."id" = "course_membership_students"."entity_role_id"
+        ) AS "num_steps_completed"
+      SQL
+    ).where(course_membership_period_id: periods.map(&:id)).group_by(&:course_membership_period_id)
```
Comment (Dantemss) [inline app/routines/push_salesforce_course_stats.rb:87]: Reduce memory usage

**Review by nathanstitt — APPROVED** [review]

**Pull request merged**